tests/matrix_multiply_tests/matrix_multiply.futil

```diff
@@ -0,0 +1,232 @@
+    @external arr_2 = comb_mem_d2(32, 4, 4, 4, 4);
+    @external mem_output = comb_mem_d2(32, 4, 4, 4, 4);
+
+    // Constants
+    const0 = std_const(32,3);
+    const1 = std_const(32,0);
+    const2 = std_const(32,1);
+
+    // Registers
+    i = std_reg(32);
+    j = std_reg(32);
+    k = std_reg(32);
+    acc = std_reg(32);
+    val_a = std_reg(32);
+    val_b = std_reg(32);
+
+    // Multiply-and-accumulate unit
+    mul_add = multiply_and_add();
+
+    // Comparators
+    le0 = std_le(32);
+    le1 = std_le(32);
+    le2 = std_le(32);
+
+    // Adder
+    add = std_add(32);
+
+    // Bit Slicer
+    slice0 = std_slice(32, 4);
+    slice1 = std_slice(32, 4);
+    slice2 = std_slice(32, 4);
+  }
+
+  wires {
+    // Loop Conditions
+    comb group cond0 { 
+      le0.left = i.out; 
+      le0.right = const0.out;
+    }
+    comb group cond1 { 
+      le1.left = j.out; 
+      le1.right = const0.out; 
+    }
+    comb group cond2 { 
+      le2.left = k.out; 
+      le2.right = const0.out; 
+    }
+
+    // Loop Initialization
+    group let0<"promotable"=1> { 
+      i.in = const1.out; 
+      i.write_en = 1'd1; 
+      let0[done] = i.done; 
+    }
+    group let1<"promotable"=1> { 
+      j.in = const1.out; 
+      j.write_en = 1'd1; 
+      let1[done] = j.done; 
+    }
+    group let2<"promotable"=1> { 
+      k.in = const1.out; 
+      k.write_en = 1'd1; 
+      let2[done] = k.done; 
+    }
+
+    // Accumulator Reset
+    group begin_acc {
+      acc.in = const1.out;
+      acc.write_en = 1'd1;
+      begin_acc[done] = acc.done;
+    }
+
+    group read_data {
+      slice0.in = i.out;
+      slice1.in = k.out;
+      slice2.in = j.out;
+      arr_1.addr0 = slice0.out; 
+      arr_1.addr1 = slice1.out;
+      arr_2.addr0 = slice1.out; 
+      arr_2.addr1 = slice2.out;
+
+      val_a.in = arr_1.read_data;
+      val_a.write_en = 1'd1;
+      val_b.in = arr_2.read_data;
+      val_b.write_en = 1'd1;
+      read_data[done] = val_b.done;
+    }
+
+    // group update_output {
+      
+    // }
+
+    // Matrix Multiplication Update
+    // group update_acc {
+    //   // mul_add.reg_done = 1'd1;
+    //   // mul_add.reg_out = acc.out;
+    //   slice0.in = i.out;
+    //   slice1.in = k.out;
+    //   arr_1.addr0 = slice0.out; 
+    //   arr_1.addr1 = slice1.out;
+    //   arr_2.addr0 = slice1.out; 
+    //   arr_2.addr1 = slice0.out;
+    //   // mul_add.a = arr_1.read_data;
+    //   // mul_add.b = arr_2.read_data;
+    //   acc.in = mul_add.reg_in;
+    //   acc.write_en = mul_add.reg_write_en;
+    //   update_acc[done] = acc.done;
+    // }
+
+    // Writing results
+    group write_result {
+      slice0.in = i.out;
+      slice1.in = j.out;
+      mem_output.addr0 = slice0.out;
+      mem_output.addr1 = slice1.out;
+      mem_output.write_data = acc.out;
+      mem_output.write_en = 1'd1;
+      write_result[done] = mem_output.done;
+    }
+
+    // Loop Increment
+    group updI<"promotable"=1> { 
```

Comment:
Small nitpick, but we use snake case for groups! (i.e. I'd change this to `upd_i`)